### ISSUE TITLE
Mixture of Experts support

### ISSUE THREAD

**Issue by ani300** [torch-spyre/torch-spyre#3565]
### What feature would you like to be added?

We should have custom operations that can be used from either hf-adapters of vLLM for the usual suspects in MoE models. For more information, see RFC https://github.com/torch-spyre/RFCs/pull/34

### Why is this feature needed?

Supporting most modern models.

---
**Related epic:** #866 (Indirect access) is the enabling capability for MoE (expert gather / scatter-combine). It owns the general indirect-access sub-issues; the MoE-specific router/gather/scatter bugs are tracked as sub-issues here.

https://github.com/torch-spyre/torch-spyre/issues/695 - This is the support epic for GPT-OSS 20B, which needs MoE.
https://github.com/torch-spyre/torch-spyre/issues/3700 - This is the support epic for the Gemma 4 family, also needs MoE.

**Comment by ani300** [comment]
## Backend blockers surfaced by the all-device Gemma-4 26B-A4B MoE (Approach A, dense chunked prefill)

While bringing up the all-device MoE FFN (full-model prefill now runs numerically-correct on-card — see hf-adapters PR #293), three backend/compiler issues each required an adapter-side workaround. Filing them as sub-issues so the workarounds can eventually be removed:

- #3768 — **topk over a pow2 stick-multiple width (64, 128)** crashes the DDC scheduler with "Incorrect chunk size". Worked around by padding router logits to a non-pow2 width (160) before topk. Distinct from #3636 (topk shape/dtype limits).
- #3769 — **>32 expert GEMM-chains in one fused sdsc program** fails the DDC scheduler (hard threshold between 32 and 64 experts, not a timeout). Worked around by chunking the expert loop into ≤32-expert compiled regions — which is what forces the host-driven chunk loop (120 dispatches/forward).
- #3770 — **non-zero storage_offset device-tensor slice as a compile input reads wrong storage**, silently (no crash). Worked around by host-slicing each chunk's expert weights then `.to("spyre")` so every compile input is offset-0.

#3768 and #3770 are pure correctness/compile blockers; #3769 is the main steady-state performance driver (an on-device expert loop or larger fused-program budget would remove the host chunking).

**Comment by AdnanHoque** [comment]
Dense all-expert compiler-path tracking is in #3928.

Current state:

```text
merged into main: #4075, #4039, #4040
open address branch: #4078 → #4041
open carried-sum branch: #4096 → #4042
integration: both open leaves on current main
```

Each open PR records its direct prerequisite and its independent responsibility.

**Comment by AdnanHoque** [comment]
Dense expert-loop compiler status:

- Parent tracker: #3928.
- Merged mechanisms: #4075, #4039, #4040, #4096, #4078.
- Active review leaf: #4041 at `6c649d46` for proven direct weight reads.
- Queued independent leaf: #4042 at `c0135099` for loop-carried terminal accumulation and final ownership verification.

Both leaves are now directly based on `main@9aa76993`. Final acceptance remains structural proof, device correctness, and a same-stack performance comparison.

**Comment by AdnanHoque** [comment]
Dense expert-loop compiler update:

- Parent tracker: #3928.
- Landed mechanisms: #4075, #4039, #4040, #4096, #4078, and #4041.
- Active and final implementation leaf: #4038/#4042 at `97fc388c`, directly on `main@3855d111`.

#4042 carries the terminal `[T,H]` sum through the counted expert loop and verifies final physical row ownership. After it lands, acceptance requires the combined structural check, device correctness, and a same-stack performance comparison.

**Comment by AdnanHoque** [comment]
Dense expert-loop compiler campaign complete and accepted:

- Parent tracker #3928 is closed.
- All mechanisms are merged: #4075, #4039, #4040, #4096, #4078, #4041, and #4042.
- Final compiler state: `main@3358f39e`.
- Structural program checks, device numerical correctness, and same-stack device-performance acceptance passed.

This completes the dense all-expert prefill compiler path within the broader MoE epic.

**Comment by AdnanHoque** [comment]
Accepted dense expert-loop performance results, with boundaries kept separate:

- Expert FFN only, chunked versus persistent: `373.1 ms → 42.6 ms` (`8.8×`).
- Full Gemma 4 26B prefill, batch size 1 and sequence length 512: `6.4 s → 1.6 s` (`4.0×`).

The first isolates the expert FFN; the second is the end-to-end model prefill result.

**Comment by ani300** [comment]
With the merge of https://github.com/torch-spyre/hf-adapters/pull/293 this is now complete
